google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/sequencer.rb

```diff
@@ -0,0 +1,103 @@
+# limitations under the License.
+
+
+require "monitor"
+
+module Google
+  module Cloud
+    module PubSub
+      class Subscriber
+        ##
+        # @private The sequencer's job is simple, keep track of all the
+        # streams's recieved message and deliver the messages with an
+        # ordering_key in the order they were recieved. The sequencer ensures
+        # only one callback can be performed at a time per ordering_key.
+        class Sequencer
+          include MonitorMixin
+
+          ##
+          # @private Implementation attributes.
+          attr_reader :keys, :messages, :process_callback
+
+          ##
+          # @private Create an empty Subscriber::Sequencer object.
+          def initialize &block
+            raise ArgumentError if block.nil?
+
+            @seq_hash = Hash.new { |hash, key| hash[key] = [] }
+            @process_callback = block
+
+            super() # to init MonitorMixin
+          end
+
+          ##
+          # @private Add a RecievedMessage to the sequencer.
+          def add message
+            # Messages without ordering_key are not managed by the sequencer
+            if message.ordering_key.empty?
+              @process_callback.call message
+              return
+            end
+
+            perform_callback = synchronize do
+              # The purpose of this block is to add the message to the
+              # sequencer, and to return whether the message should be processed
+              # immediately, or whether it will be processed later by #next. We
+              # want to ensure that these operations happen atomically.
+
+              @seq_hash[message.ordering_key].push message
+              @seq_hash[message.ordering_key].count == 1
+            end
+
+            @process_callback.call message if perform_callback
+          end
+
+          ##
+          # @private Indicate a RecievedMessage was processed, and the next in
+          # the queue can now be processed.
+          def next message
+            # Messages without ordering_key are not managed by the sequencer
+            return if message.ordering_key.empty?
+
+            next_message = synchronize do
+              # The purpose of this block is to remove the message that was
+              # processed from the sequencer, and to return the next message to
+              # be processed. We want to ensure that these operations happen
+              # atomically.
+
+              # The message should be at index 0, so this should be a very quick
+              # operation.
+              if @seq_hash[message.ordering_key].first == message
+                @seq_hash[message.ordering_key].shift
+              else
+                # This indicates something went wrong. Find the first occurrance
+                # of the message and remove it.
+                msg_index = @seq_hash[message.ordering_key].index message
+                @seq_hash[message.ordering_key].delete_at msg_index if msg_index
```

Comment:
Shouldn't we delete all subsequent messages, too? Or shut down the subscriber? And maybe we need to log an error message? Otherwise, we are going to deliver things out of order.

Reply:
We had some concurrency bugs during development and this code was used to help diagnose the problem. Once we got all the concurrency issues settled we haven't had any instances where this code was being invoked.

I don't know how the Subscriber should behave in this instance. Should it nack all the remaining messages? Or simply remove them from inventory? What should it do if it receives more messages for the `ordering_key`?

Reply:
If we are confident this won't happen, then may be we just throw an exception if it does. You are right, we can't just nack the messages currently outstanding, but would have to nack all subsequent messages. In that case, we probably are just better off treating this as a serious exception and stopping everything.

Reply:
I am confident this won't happen. I have updated this code to raise a `OrderedMessageDeliveryError` error if it does, which will stop future messages with the same ordering key from being processed by the Subscriber's callback. But like with other concurrent failures, the Subscriber attempts to recover as best as it can. (For example, if the user callback raises an error, the error is handled and indicated to the user but the Subscriber recovers and continues to process messages using the user callback.)

When an unhandled error is raised on a ruby thread, only thread is killed. Only unhandled errors raise on the main thread will kill the process. 

This error is raised on a thread from the thread pool, and not the thread that the Subscriber was created on. We use the `Subscriber#on_error` callback to notify the user when a concurrent error like this has occurred, but we aren't re-raising the error on the Subscriber's thread in an attempt to kill the process. (And there is no guarantee that the Subscriber was created on the main thread.)

We can go the extra step and also stop the subscriber when this happens, but it might be confusing for users to their Subscriber stopped without an error raised on the Subscriber's thread. So for now the implementation is using the `Subscriber#on_error` callback. Is this acceptable?

Reply:
CC @pradn